Adds ability to pass options to simpleMDE

Duder-onomy opened the pull request:
* Updates readme with examples and instructions around options
* Adds new public options parameter to the simple-mde component
* Updates simple-mde component to merge instance options into global options into default (internal) simple-mde options before creating the instance
* Adds example to the dummy app to show usage of instance options
* Adds integration test for disabling the simple-mde toolbar via options parameter

Update:
* Adds ember-assign-polyfill as travis.yml is attempting to support versions less than 2.4

Commit:
Adds ability to pass options to simpleMDE

* Updates readme with examples and instructions around options
* Adds new public options parameter to the simple-mde component
* Updates simple-mde component to merge instance options into global options into default (internal) simple-mde options before creating the instance
* Adds example to the dummy app to show usage of instance options
* Adds integration test for disabling the simple-mde toolbar via options parameter

Commit:
Adds ember-assign-polyfill
* this should be removed when support for versions < 2.4 is removed

Duder-onomy commented:
Actually, dont merge this yet. 
SimpleMDE needs its global in the namespace to assign custom toolbar support. 
If you look at the custom toolbar stuff: 
https://github.com/NextStepWebs/simplemde-markdown-editor#toolbar-icons

Im gonna make another commit now with a workaround... Looking forward to see what you think.

Commit:
Adds support to define custom toolbar options from the consuming apps config.
SimpleMDE assumes that you define the config from a JS context, in Ember our configs are defined outside our JS runtime.
Thus, we need to be able to express the simpleMDE action handlers as strings and unpack them against the SimpleMDE global.

* Adds methods to the simple-mde component to unpack toolbar options against the window if they have been expressed as strings.
* Adds documentation to support this as it is a minor deviation from raw simpleMDE implementations

Duder-onomy commented:
Ok, I added support for toolbar actions to be expressed as strings and unpacked against the window.

Commit:
Adds a semicolon.

marceloandrader commented on addon/components/simple-mde.js:
```diff
@@ -19,17 +26,69 @@ export default Ember.TextArea.extend({
+   * Thus, we unpack them and restore the global reference.
+   * If the toolbar action handler is a string, we attempt to reference the global function reference matching that string.
+   */
+  unpackToolbarOption: function(toolbarOption) {
```
@Duder-onomy  can you put some sample on how to use this, please?

wuarmin commented:
Great PR!

marceloandrader reviewed (changes requested):
@Duder-onomy  I'd like to close this PR, could you just add a sample for the toolbar functions then I will merge it.

malatin3 commented:
Any chance of getting this closed out soon? I'd love to start using this but the ability to set options is holding me back.

marceloandrader commented:
I just needed a sample for the toolbar option methods. And. I'll merge it.

On Wed, Jun 14, 2017, 15:33 Gregory Malatino <notifications@github.com>
wrote:

> Any chance of getting this closed out soon? I'd love to start using this
> but the ability to set options is holding me back.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/smith-carson/ember-simplemde/pull/5#issuecomment-308549810>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADg0BW2oOPanoyh2bRZVCZ8ynmZG1owks5sEEOFgaJpZM4NPzK7>
> .
>

Duder-onomy commented:
Just saw this. Sorry guys, I will add an example of the toolbar options right now. 
Basically, simpleMde requires you to use function references from the global `window.simpleMde`when defining toolbar click handlers, but if we want to define handlers in the ember config, then we cannot reference these functions in the same way. 

I added code to unpack these funciton references when the simpleMde component inits. So if you had, `simpleMde.getTheThing.twice` as a toolbar button click handler in your config, we will unpack that against the window like `window.simpleMde.getTheThing.twice`. Basically, you can express function references as strings in your config which will be unpacked against the window when simpleMde tries to init. 

Hope that makes sense. I will update the PR right now.

Commit:
Adds more notes to readme to attempt to explain in more detail the issue with defining toolbar action handlers in the ember config
Adds more examples to the dummy app to illustrate a few different ways to pass instance options including custom action handlers

Duder-onomy commented:
Added more notes and more examples. Please let me know if there is anything else I can do.

marceloandrader commented:
thank you @Duder-onomy  it's merged, I'll publish a new version.

marceloandrader commented:
Published in 1.1.0
